file.go

```diff
@@ -251,29 +264,45 @@ func (f *File) Close() error {
 	return errClose
 }
 
+// Readonly returns true if the file has been opened in readonly mode.
+func (f *File) Readonly() bool {
+	return f.readonly
+}
+
 // Begin creates a new read-write transaction. The transaction returned
 // does hold the Reserved Lock on the file. Use Close, Rollback, or Commit to
 // release the lock.
-func (f *File) Begin() *Tx {
+func (f *File) Begin() (*Tx, error) {
 	return f.BeginWith(TxOptions{Readonly: false})
 }
 
 // BeginReadonly creates a new readonly transaction. The transaction returned
 // does hold the Shared Lock on the file. Use Close() to release the lock.
-func (f *File) BeginReadonly() *Tx {
+func (f *File) BeginReadonly() (*Tx, error) {
 	return f.BeginWith(TxOptions{Readonly: true})
 }
 
 // BeginWith creates a new readonly or read-write transaction, with additional
 // transaction settings.
-func (f *File) BeginWith(settings TxOptions) *Tx {
+func (f *File) BeginWith(settings TxOptions) (*Tx, error) {
+	return f.beginTx(settings)
+}
+
+func (f *File) beginTx(settings TxOptions) (*Tx, reason) {
+	const op = "txfile/begin-tx"
+
+	if f.readonly && !settings.Readonly {
+		msg := "can not start writable transaction on readonly file"
+		return nil, &Error{op: op, kind: InvalidOp, ctx: f.errCtx(), msg: msg}
+	}
+
 	tracef("request new transaction (readonly: %v)\n", settings.Readonly)
 	lock := f.locks.TxLock(settings.Readonly)
 	lock.Lock()
 	tracef("init new transaction (readonly: %v)\n", settings.Readonly)
-	tx := newTx(f, lock, settings)
+	tx := newTx(f, f.txids.Inc(), lock, settings)
```

Comment:
A bit surprised that `newTx` can't return an error but checked the code and looks good.